### ISSUE TITLE
singleGroups rule warns about required parentheses

### ISSUE THREAD

**Issue by fasttime** [jshint/jshint#2548]
For example here:

    str = (1234567890).toString(16);
          ^ Unnecessary grouping operator.

Note that removing the parentheses from the line above results in a syntax error.

**Comment by lukeapage** [comment]
```
/* jshint singleGroups: true */
str = (1).toString(); // invalid warn
str = ("1").toString(); // invalid warn
str = (a).toString(); // should warn
```

**Comment by fasttime** [comment]
@lukeapage Thanks for the reply. I think that `("1").toString()` is actually the same as `"1".toString()`. So at least there the warning should be valid.

**Comment by nicolo-ribaudo** [comment]
```js
// jshint singleGroups: true
str = ({}).toString(); // invalid warn
str = (1.).toString(); // should warn
str = (1.0).toString(); // should warn
str = (1e10).toString(); // should warn
```

**Comment by fasttime** [comment]
@nicolo-ribaudo Here again for `({}).toString()` - the parentheses around `{}` are not needed, so the warning is fine. `{}.toString()` is perfectly valid JavaScript, unless we make the mistake to input it directly in a debug console (where `{}` is interpreted as an empty code block rather than an object). And in fact `alert({}.toString())` works well without parentheses around `{}`.

**Comment by jugglinmike** [comment]
Yup, it's because `{` is interpreted as a block delimiter in a statement position and an object initializer in an expression position. So far, the only invalid warning is the one originally reported: `(1).toString()`.

Self-assigning since `singleGroups` is my baby

**Comment by jugglinmike** [comment]
This bug was fixed via gh-2571 and, the fix was released in JSHint 2.9.1.
